uplink/src/base/bridge/actions_lane.rs

```diff
@@ -0,0 +1,904 @@
+use flume::{bounded, Receiver, RecvError, Sender, TrySendError};
+use log::{debug, error, info, warn};
+use serde::{Deserialize, Serialize};
+use tokio::select;
+use tokio::time::{self, Instant, Sleep};
+
+use std::collections::HashSet;
+use std::fs;
+use std::path::PathBuf;
+use std::{collections::HashMap, fmt::Debug, pin::Pin, sync::Arc, time::Duration};
+
+use super::data_lane::DataBridgeTx;
+use super::ActionBridgeShutdown;
+use crate::base::ActionRoute;
+use crate::{Action, ActionResponse, Config};
+
+const TUNSHELL_ACTION: &str = "launch_shell";
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Receiver error {0}")]
+    Recv(#[from] RecvError),
+    #[error("Io error {0}")]
+    Io(#[from] std::io::Error),
+    #[error("Serde error {0}")]
+    Serde(#[from] serde_json::Error),
+    #[error("Action receiver busy or down")]
+    UnresponsiveReceiver,
+    #[error("No route for action {0}")]
+    NoRoute(String),
+    #[error("Action timedout")]
+    ActionTimeout,
+    #[error("Another action is currently being processed")]
+    Busy,
+}
+
+struct RedirectionError(Action);
+
+pub struct ActionsBridge {
+    /// All configuration
+    config: Arc<Config>,
+    /// Tx handle to give to apps
+    status_tx: Sender<ActionResponse>,
+    /// Rx to receive action status from apps
+    status_rx: Receiver<ActionResponse>,
+    /// Actions incoming from backend
+    actions_rx: Receiver<Action>,
+    /// Handle to DataBridge
+    data_tx: DataBridgeTx,
+    /// Apps registered with the bridge
+    /// NOTE: Sometimes action_routes could overlap, the latest route
+    /// to be registered will be used in such a circumstance.
+    action_routes: HashMap<String, ActionRouter>,
+    /// Action redirections
+    action_redirections: HashMap<String, String>,
+    /// Current action that is being processed
+    current_action: Option<CurrentAction>,
+    parallel_actions: HashSet<String>,
+    ctrl_rx: Receiver<ActionBridgeShutdown>,
+    ctrl_tx: Sender<ActionBridgeShutdown>,
+    shutdown_handle: Sender<()>,
+}
+
+impl ActionsBridge {
+    pub fn new(
+        config: Arc<Config>,
+        data_tx: DataBridgeTx,
```

Comment:
There should be no reference to DataBridge in ActionsBridge

Reply:
Actions need to be responded to through the `action_status` stream, which works the same way as every other stream and is handled on the `data_lane`, which is why we have this mention.

Reply:
Does complete separation mean action status as a stream is handled completely outside of the purview of `data_lane`? Why would that be a good idea?

Reply:
Straight forwardness. Current split is to manage code better ( these 2 paths are un-necessarily coupled together affecting readability)

This PR's actions_lane is forwarding action responses to bridge (by adding handles to bridge and modifying action_response handlers) just to make it go to serializer via bridge for some reason. Rationalizing this routing, debugging actions path (there's more host to debug actions responses now) becomes difficult. Use Streams in action_lane instead to make things straight forward

Reply:
Alright, implementing the logic to ensure action status is flushed on time will require some refactoring, so expect to see this overflow into the next cycle.

Reply:
Have completed the requested change